rotkehlchen/chain/evm/decoding/merkl/decoder.py

```diff
@@ -0,0 +1,112 @@
+from rotkehlchen.globaldb.cache import (
+    globaldb_get_general_cache_values,
+)
+from rotkehlchen.globaldb.handler import GlobalDBHandler
+from rotkehlchen.history.events.structures.types import HistoryEventSubType, HistoryEventType
+from rotkehlchen.logging import RotkehlchenLogsAdapter
+from rotkehlchen.types import CacheType, ChecksumEvmAddress
+from rotkehlchen.utils.misc import bytes_to_address
+
+from .constants import CPT_MERKL, MERKL_DISTRIBUTOR_ADDRESS
+from .utils import query_merkl_rewards
+
+if TYPE_CHECKING:
+    from rotkehlchen.chain.evm.decoding.base import BaseDecoderTools
+    from rotkehlchen.chain.evm.node_inquirer import EvmNodeInquirer
+    from rotkehlchen.user_messages import MessagesAggregator
+
+logger = logging.getLogger(__name__)
+log = RotkehlchenLogsAdapter(logger)
+
+
+class MerklDecoder(DecoderInterface, ReloadableDecoderMixin):
+
+    def __init__(
+            self,
+            evm_inquirer: 'EvmNodeInquirer',
+            base_tools: 'BaseDecoderTools',
+            msg_aggregator: 'MessagesAggregator',
+    ) -> None:
+        super().__init__(
+            evm_inquirer=evm_inquirer,
+            base_tools=base_tools,
+            msg_aggregator=msg_aggregator,
+        )
+        self.rewards_protocols: defaultdict[ChecksumEvmAddress, defaultdict[ChecksumEvmAddress, list[str]]] = defaultdict(lambda: defaultdict(list))  # noqa: E501
+
+    def reload_data(self) -> Mapping['ChecksumEvmAddress', tuple[Any, ...]] | None:
+        """Check that cache is up to date and refresh cache from db.
+        Returns a fresh addresses to decoders mapping.
+        """
+        if should_update_protocol_cache(self.base.database, CacheType.MERKL_REWARD_PROTOCOLS) is True:  # noqa: E501
+            query_merkl_rewards(evm_inquirer=self.evm_inquirer, base_tools=self.base)
+        elif len(self.rewards_protocols) != 0:
+            return None  # we didn't update the globaldb cache, and we have the data already
+
+        with GlobalDBHandler().conn.read_ctx() as cursor:
+            for account in self.base.tracked_accounts.get(self.evm_inquirer.blockchain):
+                for tokens_protocols in globaldb_get_general_cache_values(
+                    cursor=cursor,
+                    key_parts=(
+                        CacheType.MERKL_REWARD_PROTOCOLS,
+                        str(self.evm_inquirer.chain_id.serialize_for_db()),
+                        account,
+                    ),
+                ):
+                    token, protocols = tokens_protocols.split('_')
+                    self.rewards_protocols[account][string_to_evm_address(token)] = protocols.split(',') if protocols else []  # noqa: E501
+
+        return self.addresses_to_decoders()
+
+    def _decode_reward_claim(self, context: DecoderContext) -> DecodingOutput:
+        """Decode a Merkl reward claim event."""
+        if context.tx_log.topics[0] != REWARD_CLAIMED:
+            return DEFAULT_DECODING_OUTPUT
+
```

Comment:
Why not change it so you keep the current logic for the decoder but instead of the cache you fetch on demand the two things that you need, the campaigns and the user rewards and then check against it. Yes, it is a remote extra query but we do something similar with other decoders already. This way only the people using merkle has to store merkl data. I checked for example my addresses and I don't have anything so I would be storing stuff that I don't use + pay the price of syncing this data for ever. Does it make sense?